### ISSUE TITLE
quaternion_to_rotation_matrix silently upcasts float16/bfloat16 to float32 (hard-coded torch.tensor(1.0))

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#3954]
`quaternion_to_rotation_matrix` silently upcasts `float16` and `bfloat16` inputs to `float32`. It is the only symbol in the rotation-representation group that does not preserve its input dtype — `normalize_quaternion` and `quaternion_to_axis_angle` on the very same input return `float16`/`bfloat16`.

The cause is a single hard-coded float32 literal (`kornia/geometry/conversions.py:699`):

```python
one: torch.Tensor = torch.tensor(1.0)

matrix_flat: torch.Tensor = torch.stack(
    (
        one - (tyy + tzz),
        ...
    ),
    dim=-1,
)
```

`torch.tensor(1.0)` has no `dtype`/`device` argument, so it is `float32` on cpu; type promotion inside `torch.stack` then lifts the whole matrix to `float32`.

## Repro

```python
import torch
from kornia.geometry.conversions import (
    normalize_quaternion as nq,
    quaternion_to_axis_angle as q2aa,
    quaternion_to_rotation_matrix as q2R,
)

for dt in (torch.float16, torch.bfloat16, torch.float32, torch.float64):
    q = torch.tensor([1.0, 0.0, 0.0, 0.0], dtype=dt)
    print(f"{str(dt):<16} q2R -> {str(q2R(q).dtype):<16} nq -> {str(nq(q).dtype):<16} q2aa -> {q2aa(q).dtype}")
```

## Observed

```
torch.float16    q2R -> torch.float32    nq -> torch.float16    q2aa -> torch.float16
torch.bfloat16   q2R -> torch.float32    nq -> torch.bfloat16   q2aa -> torch.bfloat16
torch.float32    q2R -> torch.float32    nq -> torch.float32    q2aa -> torch.float32
torch.float64    q2R -> torch.float64    nq -> torch.float64    q2aa -> torch.float64
```

`float64` survives only because `float64` wins the promotion against the `float32` literal; the two half-precision dtypes lose it.

## Expected

`quaternion_to_rotation_matrix(q).dtype == q.dtype` for every floating dtype, as every other symbol in this group already guarantees. The one-line fix is to build the literal from the input:

```python
one = torch.tensor(1.0, dtype=quaternion.dtype, device=quaternion.device)
```

or to replace it with `torch.ones_like(tyy)`.

Why it matters beyond tidiness: in a mixed-precision or half-precision pipeline this call silently doubles the memory of the rotation matrix and forces every downstream op into `float32`, either breaking `dtype` assertions or, worse, passing them quietly while defeating the precision policy the user configured. It is also the kind of defect that hides from tests parameterized only over `float32`/`float64`.

## Classification

**Broken output** in the narrow sense that the returned dtype contradicts kornia's uniform dtype-preservation contract, but with essentially zero blast radius: the numeric values are unchanged, only the container dtype differs, and the change affects exactly the two half-precision paths. Per the reference-core lifecycle (Part I / Part III §2) this is an internal repair with no observable default/semantic change on `float32`/`float64` — it can land as an ordinary bugfix under #3916-style byte-identity discipline on those paths.

## Environment

kornia 0.9.0rc1 @ `3358ef16`, torch 2.9.1, cpu. Executed from a worktree carrying the batch-3a docs/tests commits on top of `3358ef16`; `quaternion_to_rotation_matrix` is byte-identical to `origin/main` there (no diff hunk lands in it; function body compares AST-identical).

Posted on behalf of @ducha-aiki by Claude (Fable 5).

**Comment by ducha-aiki** [comment]
Correction to the scope of this issue, found while writing the convention pins for it (batch 3b of #3941).

**The upcast happens only for unbatched `(4,)` input.** Every batched shape preserves the input dtype:

```python
import torch
from kornia.geometry.conversions import quaternion_to_rotation_matrix as q2R

for dtype in (torch.float16, torch.bfloat16):
    for shape in [(4,), (1, 4), (2, 4), (3, 3, 4)]:
        q = torch.zeros(shape, dtype=dtype); q[..., 0] = 0.6; q[..., 1] = 0.8
        print(dtype, shape, "->", q2R(q).dtype)
```

```
torch.float16  (4,)     -> torch.float32
torch.float16  (1, 4)   -> torch.float16
torch.float16  (2, 4)   -> torch.float16
torch.float16  (3, 3, 4)-> torch.float16
torch.bfloat16 (4,)     -> torch.float32
torch.bfloat16 (1, 4)   -> torch.bfloat16
torch.bfloat16 (2, 4)   -> torch.bfloat16
torch.bfloat16 (3, 3, 4)-> torch.bfloat16
```

**Mechanism.** The float32 literal `one = torch.tensor(1.0)` is 0-dim. PyTorch's type promotion ranks participants by dimensionality within the same category: a tensor with dimensions outranks a 0-dim one, so for batched input the half-precision components (1-dim or higher) win and the result stays half. For unbatched `(4,)` input the quaternion components are themselves 0-dim, the 0-dim float32 literal ties with them on rank, and float32 wins the category — so the stacked matrix comes back float32.

This narrows the practical impact considerably: the reported "half-precision pipeline silently doubles its memory and then fails at the next matmul" is real, but only on the unbatched path. The title's unqualified "silently upcasts float16/bfloat16 to float32" should be read as scoped to `(4,)` input.

It does not change whether this is worth fixing — the dtype contract should not depend on whether the input is batched, and that inconsistency is arguably the sharper bug: the same call returns a different dtype for `q` and `q[None]`. A fix should make the literal follow the input dtype (or drop the literal), which fixes both paths at once.

The convention pins for this issue in the batch-3b PR use unbatched input, so they assert exactly the behavior above and stay valid; their comments have been scoped to say so.

Posted on behalf of @ducha-aiki by Claude (Fable 5).
